src/plugins/discover/public/application/main/hooks/utils/build_state_subscribe.ts

```diff
@@ -43,22 +44,45 @@ export const buildStateSubscribe =
     setDataView: DiscoverStateContainer['actions']['setDataView'];
   }) =>
   async (nextState: DiscoverAppState) => {
+    let savedSearchDataView;
     const prevState = appState.getPrevious();
     const savedSearch = savedSearchState.getState();
     if (isEqualState(prevState, nextState)) {
       addLog('[appstate] subscribe update ignored due to no changes', { prevState, nextState });
       return;
     }
     addLog('[appstate] subscribe triggered', nextState);
-    const { hideChart, interval, breakdownField, sort, index } = appState.getPrevious();
+    const { hideChart, interval, breakdownField, sort } = appState.getPrevious();
+    const nextQuery = nextState.query;
+    const currentDataView = internalState.getState().dataView;
+
+    if (isTextBasedQuery(nextQuery)) {
+      const nextDataView = await getDataViewByTextBasedQueryLang(
+        nextQuery,
+        currentDataView,
+        services
+      );
+      if (currentDataView !== nextDataView) {
+        addLog('[appstate] text based language change of data view', {
+          current: currentDataView,
+          next: nextDataView,
+          query: nextQuery,
+        });
+        setDataView(nextDataView);
+        savedSearchDataView = nextDataView;
+      }
+    }
+
     // Cast to boolean to avoid false positives when comparing
     // undefined and false, which would trigger a refetch
     const chartDisplayChanged = Boolean(nextState.hideChart) !== Boolean(hideChart);
     const chartIntervalChanged = nextState.interval !== interval;
     const breakdownFieldChanged = nextState.breakdownField !== breakdownField;
     const docTableSortChanged = !isEqual(nextState.sort, sort);
-    const dataViewChanged = !isEqual(nextState.index, index);
-    let savedSearchDataView;
+    const dataViewChanged = !isEqual(nextState.index, currentDataView?.id);
+    const queryChanged = !isEqual(nextState.query, prevState.query);
+    const filterChanged = !isEqual(nextState.filters, prevState.filters);
+
```

Comment:
adding more conditions that trigger data fetching given the removal of the subscription to the filter and query services.